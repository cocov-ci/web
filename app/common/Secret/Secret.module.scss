@import "styles/variables.module";

.base {
  display: flex;
  width: 100%;
  padding: 16px;

  .iconWrapper {
    margin-right: 10px;
  }

  .secretName {
    font-size: 15px;
    letter-spacing: 0;
  }

  .infoWrapper {
    display: flex;
    flex-direction: column;
    flex: 1;
  }

  &.active {
    .iconWrapper,
    .secretName,
    .lastUsed { color: $colorGreen; }
  }

  &.withMeta {
    padding: 14px 16px 15px;
  }

  &.divider {
    margin-bottom: 14px;
    border-bottom: 1px solid $colorWhiteDark;
  }
}
