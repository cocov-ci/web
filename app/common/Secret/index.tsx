import classNames from 'classnames'
import { Lock } from 'lucide-react'
import React from 'react'

import { inconsolata } from 'utils/fonts'

import RelativeTime from '../RelativeTime'

import styles from './Secret.module.scss'

type SecretMeta = {
  createdAt: Date
  createdBy: string
  lastUsed?: Date
}

type SecretProps = {
  children?: React.ReactNode
  className?: string
  name: string
  metadata?: SecretMeta
  showDivider?: boolean
}

const Secret = ({
  children,
  className,
  name,
  metadata,
  showDivider,
}: SecretProps) => {
  return (
    <div
      className={classNames(styles.base, className, {
        [styles.active]: !metadata || metadata.lastUsed !== undefined,
        [styles.divider]: showDivider === true,
        [styles.withMeta]: metadata !== undefined,
      })}
    >
      <div className={styles.iconWrapper}>
        <Lock height="14.44px" width="13px" />
      </div>
      <div className={styles.infoWrapper}>
        <span className={classNames(styles.secretName, inconsolata.className)}>
          {name}
        </span>
        {metadata && (
          <>
            <span className={styles.createdAt}>
              Created <RelativeTime timestamp={metadata.createdAt} /> by{' '}
              {metadata.createdBy}
            </span>
            <span className={styles.lastUsed}>
              {metadata.lastUsed ? (
                <>
                  Last used <RelativeTime timestamp={metadata.lastUsed} />
                </>
              ) : (
                <>Never used</>
              )}
            </span>
          </>
        )}
      </div>
      <div className={styles.accessoryWrapper}>{children}</div>
    </div>
  )
}

export default Secret
