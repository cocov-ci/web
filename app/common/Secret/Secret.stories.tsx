import type { Meta, StoryObj } from '@storybook/react'
import { Trash } from 'lucide-react'

import AccessoryButton from '../AccessoryButton'

import Secret from './index'

const meta: Meta<typeof Secret> = {
  title: 'Common/Secret',
  component: Secret,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Secret>

export const Simple: Story = {
  args: {
    name: 'GIT_CONFIG',
  },
}

const threeMonthsAgo = new Date(new Date().setMonth(new Date().getMonth() - 3))
const twoHoursAgo = new Date(new Date().setHours(new Date().getHours() - 2))

export const WithMeta: Story = {
  args: {
    name: 'GIT_CONFIG',
    metadata: {
      createdAt: threeMonthsAgo,
      createdBy: 'Cocov',
      lastUsed: twoHoursAgo,
    },
  },
}

export const WithMetaNeverUsed: Story = {
  args: {
    name: 'GIT_CONFIG',
    metadata: {
      createdAt: threeMonthsAgo,
      createdBy: 'Cocov',
    },
  },
}

export const WithAccessory: Story = {
  args: {
    name: 'GIT_CONFIG',
    children: (
      <AccessoryButton kind="squared">
        <Trash width="18px" />
      </AccessoryButton>
    ),
    metadata: {
      createdAt: threeMonthsAgo,
      createdBy: 'Cocov',
      lastUsed: twoHoursAgo,
    },
  },
}

export const WithDivider: Story = {
  args: {
    name: 'GIT_CONFIG',
    children: (
      <AccessoryButton kind="squared">
        <Trash width="18px" />
      </AccessoryButton>
    ),
    metadata: {
      createdAt: threeMonthsAgo,
      createdBy: 'Cocov',
      lastUsed: twoHoursAgo,
    },
    showDivider: true,
  },
}
