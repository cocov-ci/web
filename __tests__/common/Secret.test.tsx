import { render, screen } from '@testing-library/react'

import Secret from 'app/common/Secret'

import '@testing-library/jest-dom'

const threeMonthsAgo = new Date(new Date().setMonth(new Date().getMonth() - 3))
const twoHoursAgo = new Date(new Date().setHours(new Date().getHours() - 2))

describe('common/Secret', () => {
  it('renders Secret', () => {
    render(<Secret name="GIT_CONFIG" />)

    expect(screen.getByText('GIT_CONFIG')).toBeVisible()
  })

  it('renders Secret with "meta" param', () => {
    const { container } = render(
      <Secret
        metadata={{
          createdAt: threeMonthsAgo,
          createdBy: 'Cocov',
          lastUsed: twoHoursAgo,
        }}
        name="GIT_CONFIG"
      />,
    )

    expect(container.querySelector('.createdAt')?.textContent).toEqual(
      'Created 3 months ago by Cocov',
    )
    expect(container.querySelector('.lastUsed')?.textContent).toEqual(
      'Last used 2 hours ago',
    )
  })

  it('renders Secret with "meta" never used ', () => {
    render(
      <Secret
        metadata={{
          createdAt: threeMonthsAgo,
          createdBy: 'Cocov',
        }}
        name="GIT_CONFIG"
      />,
    )

    expect(screen.getByText('Never used')).toBeVisible()
  })

  it('renders Secret snapshots', () => {
    const { container } = render(<Secret name="GIT_CONFIG" />)

    expect(container).toMatchSnapshot()
  })
})
