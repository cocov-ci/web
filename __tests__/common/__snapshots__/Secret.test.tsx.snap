// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Secret renders Secret snapshots 1`] = `
<div>
  <div
    class="base active"
  >
    <div
      class="iconWrapper"
    >
      <svg
        class="lucide lucide-lock"
        fill="none"
        height="14.44px"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="13px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="11"
          rx="2"
          ry="2"
          width="18"
          x="3"
          y="11"
        />
        <path
          d="M7 11V7a5 5 0 0 1 10 0v4"
        />
      </svg>
    </div>
    <div
      class="infoWrapper"
    >
      <span
        class="secretName className"
      >
        GIT_CONFIG
      </span>
    </div>
    <div
      class="accessoryWrapper"
    />
  </div>
</div>
`;
